### PR TITLE
Use search to discover repositories based on topics

### PR DESCRIPTION
**Pull request by MindaugasLaganeckas**
# Description

We implement a faster way to discover repositories within the organization, when a list of topics is given as an input. Instead of collecting all repositories and asking about the topics each and every repository in the organization, we use github search API. In our setup with 1200 repositories, this change has given a performance improvement from 10 min per job dsl during scanning phase to 2 min or even 7 secs (depending on the number of repositories, branches, tags etc covered by the topic). The previous value of 10 mins was constant for any number of repositories, branches, tags etc for a given topic. 

@Sisha1980 has contributed to the PR with his brilliant ideas.

We truly believe that many more people can benefit from this PR.

# Submitter checklist
- [x] Link to JIRA ticket in description, if appropriate. Not appropriate.
- [x] Change is code complete and matches issue description. Done.
- [x] Automated tests have been added to exercise the changes. The current test set covers this optimization. The new input files for the tests are added. 
- [x] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.
The change affects scanning the repositories. It speeds up the process, when the topics are used.

# Reviewer checklist
- [ ] Run the changes and verify that the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

# Documentation changes
- [ ] Link to jenkins.io PR, or an explanation for why no doc changes are needed. Existing feature

# Users/aliases to notify

### CODE REVIEW

**Comment by MindaugasLaganeckas** [comment]
@bitwiseman : could you have a look at this PR? The build fails on Jenkins due to lack of disk space. `mvn clean very` works fine locally. Thank you in advance! :smile:

**Comment by leostolyarov** [comment]
Hey @ajellis1, here is the feature we are after.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java [hunk 1]
```diff
@@ -87,12 +87,7 @@
 import org.kohsuke.accmod.Restricted;
 import org.kohsuke.accmod.restrictions.DoNotUse;
 import org.kohsuke.accmod.restrictions.NoExternalUse;
-import org.kohsuke.github.GHMyself;
-import org.kohsuke.github.GHOrganization;
-import org.kohsuke.github.GHRepository;
-import org.kohsuke.github.GHUser;
-import org.kohsuke.github.GitHub;
-import org.kohsuke.github.HttpException;
+import org.kohsuke.github.*;
```
Comment (timja) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java:90]: can you revert the * import change please, and just set your IDE to not do this

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java [hunk 2]
```diff
@@ -926,7 +925,7 @@ public void visitSources(SCMSourceObserver observer) throws IOException, Interru
                 throw new AbortException(message);
             }
 
-            GitHubSCMNavigatorContext gitHubSCMNavigatorContext = new GitHubSCMNavigatorContext().withTraits(traits);
+            GitHubSCMNavigatorContext gitHubSCMNavigatorContext = new GitHubSCMNavigatorContext().withTraits(getTraits());
```
Comment (timja) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java:928]: is this change required?

File: src/test/resources/api/__files/body-search-user-repos-O8W78.json [hunk 3]
```diff
@@ -0,0 +1,189 @@
+{
```
Comment (timja) [inline src/test/resources/api/__files/body-search-user-repos-O8W78.json:1]: is there a corresponding test for these json files?

**Review by timja — COMMENTED** [review]
looks really cool, we might migrate over to this, some suggestions below

**Comment by MindaugasLaganeckas** [comment]
Thank you for the feedback! I will return to you as soon as possible. We are a bit busy at the moment in the office. :slightly_smiling_face:

**Comment by MindaugasLaganeckas** [comment]
I was too fast to resolve the last comment regarding test data and I cannot find the comment anymore. Anyhow, to answer your question, yes, the file(s) is used in these tests:
`org.jenkinsci.plugins.github_branch_source.GitHubSCMNavigatorTest#fetchRepos_BelongingToAuthenticatedUser_FilteredByTopic`
`org.jenkinsci.plugins.github_branch_source.GitHubSCMNavigatorTest#fetchRepos_BelongingToAuthenticatedUser_FilteredByTopic_RemovesAll` and ``org.jenkinsci.plugins.github_branch_source.GitHubSCMNavigatorTest#fetchRepos_BelongingToAuthenticatedUser_FilteredByMultipleTopics`. 

Here is the new expected url with the query parameters:
```
{
  "request" : {
    "urlPath" : "/search/repositories",
    "method" : "GET",
    "basicAuth": {
      "username": "git-user",
      "password": "git-secret"
    },
    "queryParameters" : {
      "q" : {
        "equalTo" : "topic:awesome org:stephenc"
      }
    }
  },
```

File: src/test/resources/api/__files/body-search-user-repos-O8W78.json (hunk 3 above)
Comment (MindaugasLaganeckas) [reply src/test/resources/api/__files/body-search-user-repos-O8W78.json:1]: I was too fast to resolve the last comment regarding test data and I cannot find the comment anymore. Anyhow, to answer your question, yes, the file(s) is used in these tests:
`org.jenkinsci.plugins.github_branch_source.GitHubSCMNavigatorTest#fetchRepos_BelongingToAuthenticatedUser_FilteredByTopic`
`org.jenkinsci.plugins.github_branch_source.GitHubSCMNavigatorTest#fetchRepos_BelongingToAuthenticatedUser_FilteredByTopic_RemovesAll` and `org.jenkinsci.plugins.github_branch_source.GitHubSCMNavigatorTest#fetchRepos_BelongingToAuthenticatedUser_FilteredByMultipleTopics`. 

Here is the new expected url with the query parameters:
```
{
  "request" : {
    "urlPath" : "/search/repositories",
    "method" : "GET",
    "basicAuth": {
      "username": "git-user",
      "password": "git-secret"
    },
    "queryParameters" : {
      "q" : {
        "equalTo" : "topic:awesome org:stephenc"
      }
    }
  },
```

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java (hunk 1 above)
Comment (MindaugasLaganeckas) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java]: Reverted.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java (hunk 2 above)
Comment (MindaugasLaganeckas) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java:934]: No, reverted.

**Comment by timja** [comment]
could you merge in master too please?

**Comment by timja** [comment]
@dwnusbaum would you be able to take a look please?

**Comment by MindaugasLaganeckas** [comment]
I have merged in master.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java [hunk 4]
```diff
@@ -930,7 +935,7 @@ public void visitSources(SCMSourceObserver observer) throws IOException, Interru
...
```
Comment (jetersen) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java:938]: Seems this was not reverted
```suggestion
            GitHubSCMNavigatorContext gitHubSCMNavigatorContext = new GitHubSCMNavigatorContext().withTraits(traits);
```

**Review by jetersen — APPROVED** [review]
LGTM

**Comment by MindaugasLaganeckas** [comment]
@timja : almost a year has past. This PR adds lots of value to the Jenkins users, but is not released to the public due to devs on this project being busy. Is it really not possible to do something about it?

**Comment by timja** [comment]
> @timja : almost a year has past. This PR adds lots of value to the Jenkins users, but is not released to the public due to devs on this project being busy. Is it really not possible to do something about it?

I have no control over this repo other than pestering maintainers to see if they will help

@car-roll can you help here?
I think @bitwiseman is away

**Comment by car-roll** [comment]
@timja I'll try and take a look here, but I would still prefer additional eyes on it if I were to merge.

**Comment by MindaugasLaganeckas** [comment]
@car-roll : @jetersen has provided the extra eyes already :)

**Comment by jetersen** [comment]
@MindaugasLaganeckas would be nice to get the build green :)

**Comment by MindaugasLaganeckas** [comment]
@jetersen : I would love that too :) Just don't know what is wrong. Google formatter reports a sever problem at line 1299 in src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java, but I do not know what is wrong with that line.

**Comment by jetersen** [comment]
You can run the google formatter through spotless.

`mvn spotless:apply`

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java [hunk 5]
```diff
@@ -1234,6 +1269,32 @@ public void visitSource(String sourceName, SCMSourceObserver observer)
     } catch (HttpException e) {
       throw new AbortException(e.getMessage());
     }
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void afterSave(@NonNull SCMNavigatorOwner owner) {
```
Comment (jetersen) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java:1276]: this seems like a failed merged 😅 as you can see this is already master

https://github.com/jenkinsci/github-branch-source-plugin/blob/bd6d7f4c0b17b500c07608bc8eb6c47413f4ebca/src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java#L1569

**Review by jetersen — CHANGES_REQUESTED** [review]

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java (hunk 5 above)
Comment (MindaugasLaganeckas) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java:1276]: Thanks a lot for the help! I will fix that merge mistake asap! 😄

Comment (MindaugasLaganeckas) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java]: @jetersen it is fixed now! Have a really great day! 🙂

**Comment by MindaugasLaganeckas** [comment]
@car-roll : the branch is green again 🎆

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java [hunk 6]
```diff
@@ -943,7 +953,7 @@ public void visitSources(SCMSourceObserver observer) throws IOException, Interru
       }
 
       GitHubSCMNavigatorContext gitHubSCMNavigatorContext =
-          new GitHubSCMNavigatorContext().withTraits(traits);
+          new GitHubSCMNavigatorContext().withTraits(getTraits());
```
Comment (car-roll) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java:956]: https://github.com/jenkinsci/github-branch-source-plugin/pull/344#discussion_r555225295

Comment (car-roll) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java:956]: nvm, looked more carefully at the code

Comment (MindaugasLaganeckas) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java:956]: So can the PR be merged? 😄

**Review by car-roll — APPROVED** [review]

**Comment by jetersen** [comment]
Congrats on getting it merged @MindaugasLaganeckas 👏

**Comment by car-roll** [comment]
FYI, release wise i'm going to wait for @bitwiseman to get back

**Comment by MindaugasLaganeckas** [comment]
@jetersen @car-roll @timja : thank you so much!!! Hopefully, we can get it released this year! 😉

**Comment by timja** [comment]
Could this be released please?

**Comment by jetersen** [comment]
Hopefully we can get a release before the end of the year 😅 

Or perhaps the plugin should consider enabling CD for the plugin: https://www.jenkins.io/doc/developer/publishing/releasing-cd/

**Comment by with-joy** [comment]
Can anyone help with this [issue](https://issues.jenkins.io/browse/JENKINS-67597?focusedCommentId=420749&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-420749) please?

Seems like related to this feature😅
